enterprise/server/remote_execution/snaploader/snaploader.go

```diff
@@ -128,29 +130,199 @@ type CacheSnapshotOptions struct {
 	WorkspaceFSPath string
 
 	// Labeled map of chunked artifacts backed by blockio.COWStore storage.
-	ChunkedFiles map[string]*ChunkedFile
+	ChunkedFiles map[string]*DynamicChunkedFile
 }
 
 type UnpackedSnapshot struct {
 	// ChunkedFiles holds any chunked files that were part of the snapshot.
-	ChunkedFiles map[string]*ChunkedFile
+	ChunkedFiles map[string]*DynamicChunkedFile
 }
 
-type ChunkedFile struct {
+// DynamicChunkedFile is a wrapper for a copy-on-write store that dynamically
+// fetches missing chunks from the cache
+type DynamicChunkedFile struct {
 	*blockio.COWStore
 
-	// digests caches any known digests for *non-dirty* chunks, keyed by offset.
-	digests map[int64]*repb.Digest
+	localCache interfaces.FileCache
+
+	// Name of the file
+	name string
+
+	// unloadedChunks tracks the digests of chunks that have not been dynamically
+	// loaded yet. Once they've been loaded, the chunks are stored in COWStore
+	unloadedChunks map[int64]*repb.Digest
+}
+
+func NewDynamicChunkedFile(store *blockio.COWStore, filecache interfaces.FileCache, name string) *DynamicChunkedFile {
+	return &DynamicChunkedFile{
+		COWStore:       store,
+		localCache:     filecache,
+		name:           name,
+		unloadedChunks: map[int64]*repb.Digest{},
+	}
+}
+
+func (cf *DynamicChunkedFile) fetchChunkIfMissing(offset int64) error {
+	chunkStartOffset := cf.ChunkStartOffset(offset)
+	if cf.ContainsOffset(chunkStartOffset) {
+		return nil
+	}
+
+	chunkDigest, ok := cf.unloadedChunks[chunkStartOffset]
+	if !ok {
+		return status.NotFoundErrorf("file does not contain an unloaded chunk with start offset %d", chunkStartOffset)
+	}
+
+	chunk, err := cf.fetchChunkFromLocalCache(chunkDigest, chunkStartOffset)
+	if err != nil {
+		// TODO(Maggie) - Fetch chunk remotely if remote snapshot sharing is on
+		//if status.IsNotFoundError(err) {
+		//}
+		return err
+	}
+
+	err = cf.AddChunk(chunkStartOffset, chunk)
+	if err != nil {
+		return status.WrapError(err, "add chunk to store")
+	}
+
+	delete(cf.unloadedChunks, chunkStartOffset)
```

Comment:
@bduffany One change I made here - If we dynamically fetch a chunk, we'll delete the digest from the memoized map. That means that even if we don't write to the chunk, we'll still have to recompute the digest. 

I thought it was cleaner this way, because otherwise the memoized map (offset -> digest) would contain some valid data and some invalid digests if we had written to that offset. Seems error-prone if someone accidentally tried to use that memoized map somewhere else without realizing it hadn't been synchronized with writes. Also seems wasteful to recompute the digest every time we write to a chunk if we wanted to keep the digest map always up-to-date